File: algorithms/2_medium/P/00050_powx_n/recursion.py
class Solution:
    def pow_recursion(self, x, n):
        if n == 0:
            return 1

        res = self.pow_recursion(x, n // 2)

        if (n & 1) == 1:
            return res * res * x

        return res * res

    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            return 1 / self.pow_recursion(x, abs(n))

        return self.pow_recursion(x, n)

File: algorithms/2_medium/P/00050_powx_n/recursion.rs
impl Solution {
    fn pow_recursion(x: f64, n: i64) -> f64 {
        if n == 0 {
            return 1.0_f64;
        }

        let res = Self::pow_recursion(x, n / 2);

        if ((n & 1) == 1) {
            res * res * x
        } else {
            res * res
        }
    }

    pub fn my_pow(x: f64, n: i32) -> f64 {
        let pow_positive = (n as i64).abs();
        let res = Self::pow_recursion(x, pow_positive);
        
        if n < 0 { 1.0_f64 / res } else { res }
    }
}
